### ISSUE TITLE
Incorrect TypeVar inference

### ISSUE THREAD

**Issue by adartaud** [microsoft/pyright#5988]
**Describe the bug**
In this example the bound value of the TypeVar seems to be inferred from the outer function call (`foo`) instead of the parameter from the init constructor. This obscures that there is an invariance error.

**Code or Screenshots**

![Capture d’écran 2023-09-20 à 18 35 16](https://github.com/microsoft/pyright/assets/7923710/89581f54-637e-4dd2-9c58-db43f3965c11)

**VS Code extension or command-line**
I am running pyright integrated into Pylance (v2023.6.40) in VSCode.

**Comment by erictraut** [comment]
Could you please paste your sample in text form? I'd prefer not to retype it from a screen shot.

**Comment by adartaud** [comment]
> Could you please paste your sample in text form? I'd prefer not to retype it from a screen shot.

```python
from typing import Generic, TypeVar

class Employee:
    ...

class Manager(Employee):
    ...

T = TypeVar("T", bound=Employee)


class MyList(Generic[T]):
    def __init__(self, o: T) -> None:
        super().__init__()


def foo(employees: MyList[Employee]):
    ...

# No error here, as expected
foo(MyList(Employee()))

ml = MyList(Manager())

# As expected we get an error since T is invariant
foo(ml)

# We should get the same error but we don't...
foo(MyList(Manager()))

# If we force the typing we get the error
foo(MyList[Manager](Manager()))

reveal_type(ml)
reveal_type(MyList(Manager()))
```

**Comment by erictraut** [comment]
Pyright's behavior is correct here. It's using [bidirectional type inference](https://microsoft.github.io/pyright/#/type-inference?id=bidirectional-type-inference-expected-types) to infer that the type of the subexpression `MyList(Manager))` is `MyList[Employee]` when you pass this subexpression as an argument to `foo`.

Not surprisingly, mypy (another Python type checker) generates the same results for your sample.

Here's a simpler example that may help you see why `foo(MyList(Manager()))` should not generate an error.

```python
def foo(x: list[float]):
    ...

x = [1, 2]
foo(x)  # Error

foo([1, 2])  # No Error
```
